examples/nlp/mlm_and_finetune_with_bert.py

```diff
@@ -0,0 +1,531 @@
+)
+
+# We have 25000 examples for testing
+x_test = encode(test_df.review.values)
+y_test = test_df.sentiment.values
+test_classifier_ds = tf.data.Dataset.from_tensor_slices((x_test, y_test)).batch(
+    flags.BATCH_SIZE
+)
+
+# Build dataset for end to end model input (will be used at the end)
+test_raw_classifier_ds = tf.data.Dataset.from_tensor_slices(
+    (test_df.review.values, y_test)
+).batch(flags.BATCH_SIZE)
+
+# Prepare data for masked language model
+x_all_review = encode(all_data.review.values)
+x_masked_train, y_masked_labels, sample_weights = get_masked_input_and_labels(
+    x_all_review
+)
+
+mlm_ds = tf.data.Dataset.from_tensor_slices(
+    (x_masked_train, y_masked_labels, sample_weights)
+)
+mlm_ds = mlm_ds.shuffle(1000).batch(flags.BATCH_SIZE)
+
+"""
+## Create BERT model (Pretraining Model) for masked language modeling
+
+We will create a BERT-like pretraining model architecture
+using the `MultiHeadAttention` layer.
+It will take token ids as inputs (including masked tokens)
+and it will predict the correct ids for the masked input tokens.
+"""
+
+
+class BERTLayer(layers.Layer):
+    def __init__(self, config, **kwargs):
+        super().__init__(**kwargs)
+
+        # Multi headed self-attention
+        self.attention = layers.MultiHeadAttention(
+            num_heads=config.NUM_HEAD,
+            key_dim=config.EMBED_DIM // config.NUM_HEAD,
+            name="multiheadattention",
+        )
+        # Feed-forward layer
+        self.ffn = keras.Sequential(
+            [
+                layers.Dense(config.FF_DIM, activation="relu"),
+                layers.Dense(config.EMBED_DIM),
+            ],
+            name="feedforwardnetwork",
+        )
+        self.dropout1 = layers.Dropout(0.1)
+        self.dropout2 = layers.Dropout(0.1)
+
+        self.layernorm1 = layers.LayerNormalization(epsilon=1e-6)
+        self.layernorm2 = layers.LayerNormalization(epsilon=1e-6)
+
+    def call(self, inputs):
+        query, key, value = inputs, inputs, inputs
+        attention_output = self.attention(inputs, inputs, inputs)
+        attention_output = self.dropout1(attention_output)
+        attention_output = self.layernorm1(inputs + attention_output)
+
+        ffn_output = self.ffn(attention_output)
+        fnn_output = self.dropout2(ffn_output)
+        sequence_output = self.layernorm2(attention_output + ffn_output)
+
+        return sequence_output
+
+
+class Encoder(layers.Layer):
+    def __init__(self, config, **kwargs):
+        super().__init__(**kwargs)
+        # Create BERTLayer
+        self.layer = [
+            BERTLayer(config, name="layer_._{}".format(i))
+            for i in range(config.NUM_LAYERS)
+        ]
+
+    def call(self, inputs):
+        outputs = inputs
+        for layer_module in self.layer:
+            outputs = layer_module(outputs)
+
+        return outputs
+
+
+# Loss metric to track masked language model loss
+loss_tracker = keras.metrics.Mean(name="mlm_loss")
```

Comment:
You can create this in the constructor. It should be managed by the model.

Reply:
I had included it in the constructor but I am facing an issue in saving `(.save('model.h5))` because it included the `Mean` Layer(loss tracker mean metric).